### PR TITLE
Week 4 and 7 Assignments

### PR DESCRIPTION
**Pull request by Seelamsetty-SaiPavan**

### CODE REVIEW

File: Week-4/Inheritance.js [hunk 1]
```diff
@@ -0,0 +1,19 @@
...
+
+// TODO: create the class Teacher and a method teach
+var Teacher = function() {
+    this.teach = function(subject){
+        console.log(this.name + " is now teaching " + subject);
+    }
+}
+Teacher.prototype = new Person();
+var him = new Teacher();
+
```
Comment (imranmd) [inline Week-4/Inheritance.js:17]: Good

File: Week-4/iterators.js [hunk 2]
```diff
@@ -0,0 +1,20 @@
...
+    next(){ 
+        this.temp = this.two; 
+        this.two = this.temp + this.one; 
+        this.one = this.temp; 
+        return {value: this.two} 
+    } 
+} 
+
+for(let I = 0 ; I < 1000; I++){ 
+       consolel.log(fibObj.next().value)
```
Comment (imranmd) [inline Week-4/iterators.js:19]: Good

File: Week-4/promise.js [hunk 3]
```diff
@@ -0,0 +1,36 @@
...
+const specialRandom = (num = 1, limit = 1000000) => {
+   // getting a random number
+   const random = Math.random() * limit;
+   // rounding it off to be divisible by num
+   const res = Math.round( random / num ) * num;
+   return res;
+};
+
+const myPromise = () => {
+    return new Promise((resolve, reject) => {
```
Comment (imranmd) [inline Week-4/promise.js:13]: This needs a custom promise.Pleas watch and re-look into this question : https://www.youtube.com/watch?v=1l4wHWQCCIc

File: Week-7/Spiral Order Matrix II.py [hunk 4]
```diff
@@ -0,0 +1,34 @@
...
+				result.append(array[row][startCol])
+			startRow+=1
+			endRow-=1
+			startCol+=1
+			endCol-=1
+    	return result
+if __name__ == '__main__':
+    array=[ [ 1, 2, 3 ], [ 4, 5, 6 ], [ 7, 8, 9 ] ]
+    res=spiralTraverse(array)
+    print(res)
```
Comment (imranmd) [inline Week-7/Spiral Order Matrix II.py:32]: Good, TC & SC is missing

File: Week-7/3sum.py [hunk 5]
```diff
@@ -0,0 +1,21 @@
...
+            e=len(nums)-1
+            while s<e:
+                sumhere=nums[i]+nums[s]+nums[e]
+                if abs(sumhere-target)<abs(res-target):
+                    res=sumhere
+                if sumhere<target:
+                    s+=1
+                else:
+                    e-=1
+        return res
```
Comment (imranmd) [inline Week-7/3sum.py:21]: Overall good.
What is the TC and SC, Any other alternatives considered? What is the trade off with other approaches?

File: Week-7/3sum.py [hunk 6]
```diff
@@ -0,0 +1,21 @@
+"""Problem 6.6 :
+ 3 sumGiven an array S of n integers, find three integers in S such that the sum 
+is closest to agiven number, target.Return the sum of the three integers.
+Assume that there will only beone solutionExample: given array 
+S = {-1 2 1 -4}, and target = 1. The sum that is closest to thetarget is 2. (-1 + 2 + 1 = 2)"""
+class Solution:
+    def threeSumClosest(self, nums: List[int], target: int) -> int:
+        nums.sort()
```
Comment (imranmd) [inline Week-7/3sum.py:8]: What is the sorting algo here and TC of this?

File: Week-7/Best time to busy and sell stock.py [hunk 7]
```diff
@@ -0,0 +1,20 @@
...
+     1 is not allowed because you must buy before you sell. Example 2: Input: prices =[7,6,4,3,1]
+      Output: 0 Explanation: In this case, no transactions are done and the maxprofit = 0
+       Constraints: 1 <= prices.length <= 105 0 <= prices[i] <= 104"""
+
+class Solution:
+    def maxProfit(self,prices):
+        min_price=float("inf")
+        max_profit = 0
+        for price in prices:
+            min_price=min(min_price,price)
```
Comment (imranmd) [inline Week-7/Best time to busy and sell stock.py:18]: TC & SC

Comment (imranmd) [reply Week-7/Best time to busy and sell stock.py:18]: Edge cases handling

File: Week-7/3sum.py (hunk 6 above)
Comment (imranmd) [reply Week-7/3sum.py:8]: Will this break if nums is not having any elements

File: Week-7/sort 0 1 2.py [hunk 8]
```diff
@@ -0,0 +1,28 @@
...
+sort the array in ascendingorder. 
+Example 1:Input:N = 5arr[]= {0 2 1 2 0}
+Output:0 0 1 2 2
+Explanation: 0's 1's and 2's are segregated into ascending order.
+Example 2:Input:N = 3arr[] = {0 1 0}
+Output:0 0 1Explanation: 0s 1s and 2s are segregated into ascending order.
+Time Complexity: O(N)Expected Auxiliary Space: O(1)Constraints: 1 <= N <= 10^6 0 <= A[i] <= 2"""
+def sortNumbers(A):
+    left=current=0
+    right = len(A) - 1
```
Comment (imranmd) [inline Week-7/sort 0 1 2.py:12]: Edge cases

File: Week-7/max sum contiguous sub array.py [hunk 9]
```diff
@@ -0,0 +1,24 @@
...
+    max_so_far = a[0]
+    max_ending_here = 0
+     
+    for i in range(0, size):
+        max_ending_here = max_ending_here + a[i]
+        if max_ending_here < 0:
+            max_ending_here = 0
+        elif (max_so_far < max_ending_here):
+            max_so_far = max_ending_here
+
```
Comment (imranmd) [inline Week-7/max sum contiguous sub array.py:23]: Good

**Review by imranmd — APPROVED** [review]
Overall good work. Approving with suggestions.

1. Edge cases are missing
2. Need to mention SC & TC for every problem, also mention sample input and output wherever required
3. Any other approaches , please mention and what was the trade off factor
4. Custom promise needs to be implemented.

File: Week-4/promise.js (hunk 3 above)
Comment (Seelamsetty-SaiPavan) [reply Week-4/promise.js:13]: sure

**Pull request merged**